runners: promote --no-load argument into common code and add support in stlink_gdbrunner

mathieuchopstm opened the pull request:
Promote the existing `--no-load` argument from OpenOCD/Intel Cyclone V runners into common runner code. This argument inhibits loading of image on target before a debug session, allowing `west debug` to effectively be used as a `west attach`-after-reset.

Also add support of this argument to the `stlink_gdbserver` runner.

marc-hb reviewed (commented):
The formatting fixes belong to a separate commit

https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#multiple-commits-on-a-single-pr

> This flag functions exactly like its identically-named counterpart in the OpenOCD runner: when present, the "load" command is not added to the GDB client's command line.

... duplication which points at the lack of a common function or class. You probably not interested in changing and testing the OpenOCD runner, but you can still create that common function / class and start using it only in `stlink_gdbserver.py`. Then it's ready for both existing and future runners to use.


EDIT: as a pure coincidence, here is a current example of moving generic code to generic places, away from product-specific code:
- #98443

pdgendt commented on scripts/west_commands/runners/stlink_gdbserver.py:
```diff
@@ -110,10 +110,23 @@ def do_add_parser(cls, parser: argparse.ArgumentParser):
+            "--no-load",
+            action='store_true',
+            help="Don't flash binary before debug (no GDB 'load' command)",
+        )
```
Turn this into a `--load/--no-load` option, consistent with `--rebuild/--no-rebuild` and others. Similar to `--swd` in https://github.com/zephyrproject-rtos/zephyr/pull/98670

```python
        # Additional arguments
        parser.add_argument(
            "--load",
            default=True,
            action=argparse.BooleanOptionalAction,
            help="Optionally flash binary before debug (GDB 'load' command)",
        )
```

This is to make the arguments more consistent and let the last passed argument win.

I also agree with @marc-hb, it would be nice to turn this into a runner capability with some common handling.

mathieuchopstm commented on scripts/west_commands/runners/stlink_gdbserver.py:
```diff
@@ -110,10 +110,23 @@ def do_add_parser(cls, parser: argparse.ArgumentParser):
+            args.dev_id,
+            args.port_number,
+            args.extload,
+            args.no_load,
```
> The formatting fixes belong to a separate commit
> 
> https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#multiple-commits-on-a-single-pr

Are you referring to this? This isn't merely a "formatting" fix - `ruff` will yell at me because line is too long (arguably, I didn't check if it *does* yell at me here, but I know it will because another patch will be merged in the runner beforehand)

mathieuchopstm replied:
The problem is that this is *currently* accepted by `ruff`, but the commit adds a new argument which makes it go over the length limit and then `ruff` starts complaining. Running `ruff` beforehand will not change anything - the file is already compliant.

mathieuchopstm committed:
scripts: runners: generalize --no-load argument

Move the existing "--no-load" argument from the OpenOCD and Intel Cyclone V
runners into an argument in the runners base class such that it can be used
by all runners. Also update the existing runners to work with the common
option instead of their own.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
scripts: runners: stlink_gdbserver: add support for --no-load

Add support for the --load/--no-load argument from the base runner class.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

pdgendt reviewed (dismissed):
Nice, thanks!

mathieuchopstm commented:
Rebased to resolve conflict

marc-hb reviewed (commented):
I'm confused by this new `debug_load` capability: what happens when a future runner does _not_ to have the capability? Does that runner `-ex load` or not? Does it depend on the specific runner? Inconsistently...

- Developers could assume "loading" is a capability. No capability = no load command issued.
- But they could assume instead that _skipping_ the load is the capability. So, loading when no skip capability (similar to right now before this PR)

Generally speaking, having "two levels" of booleans adds complexity. So you could just drop the capability and force all runners to implement the option. It requires only one line of code in the runner. If some future runners can't support one of the two choices (cause they don't use `gdb`?), they could just fail when someone tries to unsupported choice.

Except... there isn't really a way to force runners to look at `args.load`, is there?

This is surprisingly more complicated that I thought.

mathieuchopstm commented:
> I'm confused by this new `debug_load` capability: what happens when a future runner does _not_ to have the capability? Does that runner `-ex load` or not? Does it depend on the specific runner? Inconsistently...

I do agree the capability name is not great. Maybe something like `noload` / `noload_debug` / `debug_without_load` would be more understandable?

> * Developers could assume "loading" is a capability. No capability = no load command issued.

Per "specification", `west debug` is *expected* to flash before starting the session:

>```
>$ west debug --help
>usage: west debug [-h] [-d DIR] [-r RUNNER] [--domain DOMAIN] [--skip-rebuild | --rebuild | --no-rebuild] [-H] [--board-dir DIR] [--gdb GDB] [--openocd OPENOCD]
>                   [--openocd-search DIR]
>
>Connect to the board, flash the program, and start a debugging session. Use "west attach" instead to attach a debugger without reflashing.
>```

Emphasis on ***`flash the program`***, hence the comment in my code:
> `# by default, 'west debug' is expected to flash before starting the session`

(to clarify again, the reason why `west flash --no-load` would be useful is that `west attach` merely starts a debugger, connects to and halts the target, whereas `west debug` also resets the target, breaks on entrypoint, ...)

> Generally speaking, having "two levels" of booleans adds complexity. So you could just drop the capability and force all runners to implement the option. It requires only one line of code in the runner. If some future runners can't support one of the two choices (cause they don't use `gdb`?), they could just fail when someone tries to unsupported choice.

What I had in mind was instead tools which would (*forcefully!*) download an image before starting the GDB server, or tools which don't use GDB at all. Looking at tree for runners which support `debug`, most do currently merely append `-ex load` to GDB commandline, but there are some exceptions:

- `Synopsys Metaware Debugger(mdb)`: seems to be its own thing entirely (i.e., not GDB)
https://github.com/zephyrproject-rtos/zephyr/blob/8c264d30ea1b5c218a2efd4288820814ff16f1e2/scripts/west_commands/runners/mdb.py#L75-L82
- `native`: executes native binary - `load`/`no-load` wouldn't apply
- `probe-rs`: uses an opaque `gdb` argument for which I can't find documentation :man_shrugging:
https://github.com/zephyrproject-rtos/zephyr/blob/8c264d30ea1b5c218a2efd4288820814ff16f1e2/scripts/west_commands/runners/probe_rs.py#L167-L187
- `spi_burn`: doesn't seem to download image when `west debug`'ing?
- `trace32`: seems to be its own thing?
https://github.com/zephyrproject-rtos/zephyr/blob/8c264d30ea1b5c218a2efd4288820814ff16f1e2/scripts/west_commands/runners/trace32.py#L131-L142

> Except... there isn't really a way to force runners to look at `args.load`, is there?

Yes, hence why it should be a capability if we follow what's already in tree.

marc-hb commented:
Or, you could call it `debug_skip_load` if you hate negative capabilities even more than I do.

```python
parser.add_argument('--load', action=argparse.BooleanOptionalAction,
        help=("load image on target before 'west debug' session"
       if caps.debug_skip_load else argparse.SUPPRESS),
            default=True)
